crate2nix/src/prefetch.rs

```diff
@@ -97,62 +104,87 @@ pub fn prefetch(
         })
         .collect();
 
-    let without_hash_num = prefetchable_sources
+    let num_crates_without_hash = prefetchable_sources
         .iter()
         .filter(|SourcePrefetchBundle { hash, .. }| hash.is_none())
         .count();
 
-    let mut idx = 1;
-    for SourcePrefetchBundle {
-        source,
-        packages,
-        hash,
-    } in prefetchable_sources.iter_mut()
-    {
-        let (sha256, hash_source) = if let Some(HashWithSource { sha256, source }) = hash {
-            (sha256.trim().to_string(), *source)
-        } else {
-            eprintln!(
-                "Prefetching {:>4}/{}: {}",
-                idx,
-                without_hash_num,
-                source.to_string()
-            );
-            idx += 1;
-            (source.prefetch()?, HashSource::Prefetched)
-        };
-
-        for package in packages.iter_mut() {
-            package.source = package.source.with_sha256(sha256.clone());
-            if hash_source == HashSource::Prefetched {
-                hashes.insert(package.package_id.clone(), sha256.clone());
+    let raw_progress_bar = ProgressBar::new(num_crates_without_hash.try_into()?);
+    raw_progress_bar.set_style(
+        ProgressStyle::default_bar()
+            .template("[{elapsed_precise}] [{bar:40.cyan/blue}] {pos}/{len} ({eta})")
+            .progress_chars("#>-"),
+    );
+    let progress_bar = Arc::new(raw_progress_bar);
+    let tasks = prefetchable_sources.into_iter().map(
+        |SourcePrefetchBundle {
+             source,
+             packages,
+             hash,
+         }| {
+            let pb = progress_bar.clone();
+            async move {
+                let (sha256, hash_source) = if let Some(HashWithSource { sha256, source }) = hash {
+                    (sha256.trim().to_string(), source)
+                } else {
+                    let prefetched_hash = source.prefetch().await?;
+                    pb.inc(1);
+                    (prefetched_hash, HashSource::Prefetched)
+                };
+
+                Result::<_, Error>::Ok(packages.iter_mut().map(move |package| {
+                    let package_id = package.package_id.clone();
+                    let package_source = package.source.with_sha256(sha256.clone());
+                    (
+                        package,
+                        package_source,
+                        if hash_source == HashSource::Prefetched {
+                            Some((package_id, sha256.clone()))
+                        } else {
+                            None
+                        },
+                    )
+                }))
             }
+        },
+    );
+
+    // TODO: Is there a better way to choose this number?
+    let n_concurrent_tasks = num_cpus::get() * 10;
```

Comment:
This does not really depend on the CPUs -- more on network bandwidth and the speed of the server at the other end. Therefore, this seems to be excessive. I don't think that 160 is a reasonable number of parallel prefetches, yet this can happen with this auto detection.

I would rather go with "4" by default or something like that and make it configurable via command line options.

Reply:
I agree that using num cpus here doesn't make sense, and configuration is a good idea. However, I think we still want a relatively large number of tasks by default (tens of tasks).

If one download is slow then we can still make progress on others (of course, only up to a point).

160 might not be reasonable, but I think eg 60 would be.

I can measure roughly where the point of diminishing returns is (recognizing that this will be dependent mostly on network factors)